File: _citybox_src/apps/marketplace/android/app/src/main/java/com/citybox/data/api/ApiModels.kt
package com.citybox.data.api

import kotlinx.serialization.Serializable

/**
 * Data classes @Serializable espelhando o contrato do BFF
 * (fonte de verdade: apps/marketplace/web/src/api/types.ts).
 * Envelope padrão: `{ data, meta?, errors?[{code,message,field}] }`, camelCase.
 */

@Serializable
data class ApiEnvelope<T>(
    val data: T? = null,
    val meta: ApiPageMeta? = null,
    val errors: List<ApiErrorItem>? = null,
)

@Serializable
data class ApiErrorItem(
    val code: String? = null,
    val message: String? = null,
    val field: String? = null,
)

@Serializable
data class ApiPageMeta(
    val page: Int? = null,
    val pageSize: Int? = null,
    val total: Int? = null,
)

// ── Auth & usuário ─────────────────────────────────────────────────────────

@Serializable
data class ApiUser(
    val id: String? = null,
    val name: String = "",
    val email: String = "",
    val phone: String = "",
    val avatarUrl: String? = null,
    val avatarInitial: String = "",
    val isPlus: Boolean = false,
    val hasSeenOnboarding: Boolean? = null,
)

@Serializable
data class ApiAuthData(
    val accessToken: String = "",
    val refreshToken: String? = null,
    val expiresIn: Long? = null,
    val user: ApiUser = ApiUser(),
)

@Serializable
data class ApiSessionData(
    val user: ApiUser = ApiUser(),
    val accessToken: String? = null,
    val isAuthenticated: Boolean? = null,
)

@Serializable
data class ApiRefreshData(
    val accessToken: String = "",
    val expiresIn: Long? = null,
)

@Serializable
data class ApiUserWrapper(val user: ApiUser = ApiUser())

@Serializable
data class ApiMessageData(val message: String? = null, val sent: Boolean? = null)

// ── Catálogo ───────────────────────────────────────────────────────────────

@Serializable
data class ApiProduct(
    val id: String = "",
    val name: String = "",
    val imageUrl: String = "",
    val price: Double = 0.0,
    val originalPrice: Double? = null,
    val discountPercent: Int? = null,
    val rating: Double = 0.0,
    val reviewCount: Int = 0,
    val isFreeShipping: Boolean = false,
    val isExpress: Boolean = false,
    val category: String = "",
    val categoryId: String? = null,
    val brand: String? = null,
    val specs: Map<String, String>? = null,
)

@Serializable
data class ApiCategory(
    val id: String = "",
    val name: String = "",
    val icon: String = "",
    val colorHex: String = "",
)

@Serializable
data class ApiHomeSection(
    val id: String = "",
    val title: String = "",
    val productIds: List<String> = emptyList(),
)

@Serializable
data class ApiHomeData(
    val sections: List<ApiHomeSection> = emptyList(),
    val products: List<ApiProduct> = emptyList(),
)

@Serializable
data class ApiCategoriesWrapper(val categories: List<ApiCategory> = emptyList())

@Serializable
data class ApiCategoryProductsData(
    val category: ApiCategory? = null,
    val products: List<ApiProduct> = emptyList(),
)

@Serializable
data class ApiProductsWrapper(val products: List<ApiProduct> = emptyList())

@Serializable
data class ApiProductWrapper(val product: ApiProduct = ApiProduct())

@Serializable
data class ApiSuggestionsData(
    val suggestions: List<String> = emptyList(),
    val brands: List<String> = emptyList(),
)

@Serializable
data class ApiReview(
    val id: String = "",
    val productId: String = "",
    val author: String = "",
    val rating: Int = 0,
    val date: String = "",
    val text: String = "",
    val photoUrls: List<String>? = null,
)

@Serializable
data class ApiReviewsListData(
    val averageRating: Double = 0.0,
    val totalCount: Int = 0,
    val distribution: Map<String, Int>? = null,
    val reviews: List<ApiReview> = emptyList(),
)

@Serializable
data class ApiReviewWrapper(val review: ApiReview = ApiReview())

@Serializable
data class ApiSearchHistoryData(val queries: List<String> = emptyList())

// ── Carrinho / favoritos ───────────────────────────────────────────────────

@Serializable
data class ApiCartItem(
    val productId: String = "",
    val quantity: Int = 0,
    val product: ApiProduct? = null,
)

@Serializable
data class ApiCart(
    val items: List<ApiCartItem> = emptyList(),
    val itemCount: Int = 0,
    val subtotal: Double = 0.0,
)

@Serializable
data class ApiCartWrapper(val cart: ApiCart = ApiCart())

@Serializable
data class ApiFavoritesData(
    val productIds: List<String> = emptyList(),
    val products: List<ApiProduct> = emptyList(),
)

@Serializable
data class ApiAppliedCoupon(
    val code: String = "",
    val type: String = "FIXED",
    val value: Double = 0.0,
    val discountAmount: Double = 0.0,
)

@Serializable
data class ApiCoupon(
    val code: String = "",
    val description: String = "",
    val type: String = "FIXED",
    val value: Double = 0.0,
    val expiry: String = "",
    val isApplicable: Boolean? = null,
    val reason: String? = null,
)

@Serializable
data class ApiCouponsWrapper(val coupons: List<ApiCoupon> = emptyList())

@Serializable
data class ApiApplyCouponData(
    val appliedCoupon: ApiAppliedCoupon? = null,
    val cart: ApiCart? = null,
    val preview: ApiCheckoutPreview? = null,
)

// ── Checkout ───────────────────────────────────────────────────────────────

@Serializable
data class ApiCheckoutSession(
    val selectedAddressId: String? = null,
    val shippingOptionId: String? = null,
    val appliedCoupon: ApiAppliedCoupon? = null,
    val paymentType: String? = null,
    val paymentMethodId: String? = null,
    val boletoCpf: String? = null,
    val canConfirm: Boolean? = null,
)

@Serializable
data class ApiCheckoutPreview(
    val subtotal: Double = 0.0,
    val shipping: Double = 0.0,
    val couponDiscount: Double = 0.0,
    val pixDiscount: Double = 0.0,
    val total: Double = 0.0,
    val pixDiscountPercent: Double? = null,
    val canConfirm: Boolean? = null,
    val validationErrors: List<String>? = null,
)

@Serializable
data class ApiCheckoutSessionData(
    val cart: ApiCart = ApiCart(),
    val session: ApiCheckoutSession = ApiCheckoutSession(),
    val preview: ApiCheckoutPreview = ApiCheckoutPreview(),
)

@Serializable
data class ApiShippingOption(
    val id: String = "",
    val name: String = "",
    val deliveryEstimate: String = "",
    val price: Double = 0.0,
    val isExpress: Boolean? = null,
)

@Serializable
data class ApiShippingOptionsData(
    val options: List<ApiShippingOption> = emptyList(),
    val freeShippingMessage: String? = null,
)

@Serializable
data class ApiPaymentInput(
    val type: String,
    val paymentMethodId: String? = null,
    val cpf: String? = null,
)

@Serializable
data class ApiPaymentResult(
    val type: String = "PIX",
    val status: String = "PENDING",
    val paymentMethodId: String? = null,
    val displayName: String? = null,
    val pixCopyPaste: String? = null,
    val digitableLine: String? = null,
    val dueDate: String? = null,
)

@Serializable
data class ApiCreateOrderData(
    val order: ApiOrder = ApiOrder(),
    val payment: ApiPaymentResult? = null,
)

// ── Pedidos ────────────────────────────────────────────────────────────────

@Serializable
data class ApiOrderItem(
    val productId: String = "",
    val product: ApiProduct? = null,
    val quantity: Int = 1,
    val unitPrice: Double = 0.0,
    val subtotal: Double = 0.0,
)

@Serializable
data class ApiOrderStatusEntry(
    val status: String = "CONFIRMED",
    val date: String = "",
    val location: String? = null,
)

@Serializable
data class ApiOrderPaymentMethod(
    val type: String = "PIX",
    val displayName: String = "",
    val label: String? = null,
)

@Serializable
data class ApiOrder(
    val id: String = "",
    val items: List<ApiOrderItem> = emptyList(),
    val status: String = "CONFIRMED",
    val deliveryDate: String = "",
    val address: ApiAddress? = null,
    val paymentMethod: ApiOrderPaymentMethod? = null,
    val subtotal: Double = 0.0,
    val shipping: Double = 0.0,
    val discount: Double = 0.0,
    val pixDiscount: Double? = null,
    val total: Double = 0.0,
    val trackingCode: String? = null,
    val carrier: String? = null,
    val statusHistory: List<ApiOrderStatusEntry> = emptyList(),
    val createdAt: String? = null,
)

@Serializable
data class ApiOrderWrapper(val order: ApiOrder = ApiOrder())

@Serializable
data class ApiOrdersData(val orders: List<ApiOrder> = emptyList())

@Serializable
data class ApiTrackingEntry(
    val status: String = "CONFIRMED",
    val date: String = "",
    val location: String? = null,
    val description: String? = null,
)

@Serializable
data class ApiTracking(
    val orderId: String = "",
    val trackingCode: String = "",
    val carrier: String = "",
    val currentStatus: String = "CONFIRMED",
    val estimatedDelivery: String = "",
    val timeline: List<ApiTrackingEntry> = emptyList(),
)

@Serializable
data class ApiReturnCreatedData(
    val returnId: String = "",
    val status: String = "",
    val message: String? = null,
)

// ── Endereços / pagamentos ─────────────────────────────────────────────────

@Serializable
data class ApiAddress(
    val id: String = "",
    val label: String = "",
    val zipCode: String = "",
    val street: String = "",
    val number: String = "",
    val complement: String? = null,
    val neighborhood: String = "",
    val city: String = "",
    val state: String = "",
    val isDefault: Boolean = false,
)

@Serializable
data class ApiAddressInput(
    val label: String,
    val zipCode: String,
    val street: String,
    val number: String,
    val complement: String? = null,
    val neighborhood: String,
    val city: String,
    val state: String,
    val isDefault: Boolean = false,
)

@Serializable
data class ApiAddressesWrapper(val addresses: List<ApiAddress> = emptyList())

@Serializable
data class ApiAddressWrapper(val address: ApiAddress = ApiAddress())

@Serializable
data class ApiZipLookup(
    val zipCode: String = "",
    val street: String = "",
    val neighborhood: String = "",
    val city: String = "",
    val state: String = "",
)

@Serializable
data class ApiPaymentMethod(
    val id: String = "",
    val brand: String = "UNKNOWN",
    val lastFour: String = "",
    val expiry: String = "",
    val holderName: String = "",
    val label: String = "",
    val isDefault: Boolean = false,
)

@Serializable
data class ApiPaymentMethodsWrapper(val paymentMethods: List<ApiPaymentMethod> = emptyList())

@Serializable
data class ApiPaymentMethodWrapper(val paymentMethod: ApiPaymentMethod = ApiPaymentMethod())

@Serializable
data class ApiCardInput(
    val number: String,
    val holderName: String,
    val expiry: String,
    val cvv: String,
    val label: String? = null,
    val isDefault: Boolean = false,
)

// ── Conta / engajamento ────────────────────────────────────────────────────

@Serializable
data class ApiSettings(
    val pushOrdersEnabled: Boolean = true,
    val pushPromoEnabled: Boolean = true,
    val emailPromoEnabled: Boolean = false,
    val darkTheme: Boolean = false,
    val language: String = "pt-BR",
)

@Serializable
data class ApiSubscription(
    val isActive: Boolean = false,
    val planName: String = "",
    val priceMonthly: Double = 0.0,
    val renewalDate: String = "",
    val benefits: List<String> = emptyList(),
)

@Serializable
data class ApiNotification(
    val id: String = "",
    val type: String = "SYSTEM",
    val title: String = "",
    val body: String = "",
    val date: String = "",
    val isRead: Boolean = false,
    val deepLink: String? = null,
)

@Serializable
data class ApiNotificationsData(
    val notifications: List<ApiNotification> = emptyList(),
    val unreadCount: Int = 0,
)

@Serializable
data class ApiFaqItem(val question: String = "", val answer: String = "")

@Serializable
data class ApiFaqData(val topics: List<ApiFaqItem> = emptyList())

@Serializable
data class ApiChatMessage(
    val id: String = "",
    val text: String = "",
    val isAgent: Boolean = false,
    val time: String = "",
)

@Serializable
data class ApiChatMessagesData(val messages: List<ApiChatMessage> = emptyList())

@Serializable
data class ApiChatSendData(
    val userMessage: ApiChatMessage? = null,
    val agentMessage: ApiChatMessage? = null,
)

@Serializable
data class ApiTicket(
    val ticketId: String = "",
    val status: String = "OPEN",
    val subject: String? = null,
    val message: String? = null,
    val orderId: String? = null,
    val createdAt: String? = null,
)

@Serializable
data class ApiTicketsData(val tickets: List<ApiTicket> = emptyList())

@Serializable
data class ApiTicketCreatedData(val ticketId: String = "", val status: String = "OPEN")

// ── Conteúdo ───────────────────────────────────────────────────────────────

@Serializable
data class ApiStaticPage(
    val slug: String = "",
    val title: String = "",
    val content: String = "",
    val updatedAt: String? = null,
)

@Serializable
data class ApiBannerAction(val type: String = "", val query: String? = null)

@Serializable
data class ApiBanner(
    val id: String = "",
    val title: String = "",
    val subtitle: String? = null,
    val imageUrl: String = "",
    val action: ApiBannerAction? = null,
)

@Serializable
data class ApiBannersData(val banners: List<ApiBanner> = emptyList())

File: _citybox_src/apps/marketplace/ios/Sources/Data/ApiModels.swift
import Foundation

// MARK: - DTOs do contrato do BFF (camelCase, espelham web/src/api/types.ts)

struct ApiProduct: Codable {
    let id: String
    let name: String
    let imageUrl: String
    let price: Double
    let originalPrice: Double?
    let discountPercent: Int?
    let rating: Double
    let reviewCount: Int
    let isFreeShipping: Bool
    let isExpress: Bool
    let category: String
    let categoryId: String?
    let brand: String?
    let specs: [String: String]?
}

struct ApiCategory: Codable {
    let id: String
    let name: String
    let icon: String
    let colorHex: String
}

struct ApiHomeSection: Codable {
    let id: String
    let title: String
    let productIds: [String]
}

struct ApiHomeData: Codable {
    let sections: [ApiHomeSection]
    let products: [ApiProduct]
}

struct ApiUser: Codable {
    let id: String?
    let name: String
    let email: String
    let phone: String
    let avatarUrl: String?
    let avatarInitial: String?
    let isPlus: Bool?
    let hasSeenOnboarding: Bool?
}

struct ApiAuthData: Codable {
    let accessToken: String
    let refreshToken: String?
    let expiresIn: Int?
    let user: ApiUser
}

struct ApiSessionData: Codable {
    let user: ApiUser
    let isAuthenticated: Bool?
}

struct ApiCartItem: Codable {
    let productId: String
    let quantity: Int
    let product: ApiProduct?
}

struct ApiCart: Codable {
    let items: [ApiCartItem]
    let itemCount: Int?
    let subtotal: Double?
}

struct ApiFavoritesData: Codable {
    let productIds: [String]
    let products: [ApiProduct]?
}

struct ApiAddress: Codable {
    let id: String
    let label: String
    let zipCode: String
    let street: String
    let number: String
    let complement: String?
    let neighborhood: String
    let city: String
    let state: String
    let isDefault: Bool
}

struct ApiAddressInput: Encodable {
    let label: String
    let zipCode: String
    let street: String
    let number: String
    let complement: String?
    let neighborhood: String
    let city: String
    let state: String
    let isDefault: Bool
}

struct ApiFaqItem: Codable {
    let question: String
    let answer: String
}

struct ApiOrderItem: Codable {
    let productId: String
    let product: ApiProduct?
    let quantity: Int
    let unitPrice: Double?
    let subtotal: Double?
}

struct ApiOrderStatusEntry: Codable {
    let status: String
    let date: String
    let location: String?
}

struct ApiOrderPaymentMethod: Codable {
    let type: String
    let displayName: String
    let label: String?
}

struct ApiOrder: Codable {
    let id: String
    let items: [ApiOrderItem]
    let status: String
    let deliveryDate: String?
    let address: ApiAddress?
    let paymentMethod: ApiOrderPaymentMethod?
    let subtotal: Double
    let shipping: Double
    let discount: Double
    let pixDiscount: Double?
    let total: Double
    let trackingCode: String?
    let carrier: String?
    let statusHistory: [ApiOrderStatusEntry]?
    let createdAt: String?
}

struct ApiTrackingTimelineEntry: Codable {
    let status: String
    let date: String
    let location: String?
    let description: String?
}

struct ApiTracking: Codable {
    let orderId: String
    let trackingCode: String
    let carrier: String
    let currentStatus: String
    let estimatedDelivery: String
    let timeline: [ApiTrackingTimelineEntry]
}

struct ApiNotification: Codable {
    let id: String
    let type: String
    let title: String
    let body: String
    let date: String
    let isRead: Bool
    let deepLink: String?
}

struct ApiChatMessage: Codable {
    let id: String
    let text: String
    let isAgent: Bool
    let time: String
}

struct ApiCoupon: Codable {
    let code: String
    let description: String
    let type: String
    let value: Double
    let expiry: String
    let isApplicable: Bool?
}

struct ApiAppliedCoupon: Codable {
    let code: String
    let type: String
    let value: Double
    let discountAmount: Double?
}

struct ApiPaymentMethod: Codable {
    let id: String
    let brand: String
    let lastFour: String
    let expiry: String
    let holderName: String
    let label: String
    let isDefault: Bool
}

struct ApiShippingOption: Codable {
    let id: String
    let name: String
    let deliveryEstimate: String
    let price: Double
    let isExpress: Bool?
}

struct ApiReview: Codable {
    let id: String
    let productId: String
    let author: String
    let rating: Int
    let date: String
    let text: String
    let photoUrls: [String]?
}

struct ApiReviewsListData: Codable {
    let averageRating: Double
    let totalCount: Int
    let reviews: [ApiReview]
}

struct ApiSubscription: Codable {
    let isActive: Bool
    let planName: String
    let priceMonthly: Double
    let renewalDate: String
    let benefits: [String]
}

struct ApiSettings: Codable {
    let pushOrdersEnabled: Bool
    let pushPromoEnabled: Bool
    let emailPromoEnabled: Bool
    let darkTheme: Bool
    let language: String
}

struct ApiCheckoutSession: Codable {
    let selectedAddressId: String?
    let shippingOptionId: String?
    let appliedCoupon: ApiAppliedCoupon?
    let paymentType: String?
    let paymentMethodId: String?
    let boletoCpf: String?
    let canConfirm: Bool?
}

struct ApiCheckoutPreview: Codable {
    let subtotal: Double
    let shipping: Double
    let couponDiscount: Double
    let pixDiscount: Double
    let total: Double
    let canConfirm: Bool?
    let validationErrors: [String]?
}

struct ApiCheckoutSessionData: Codable {
    let cart: ApiCart
    let session: ApiCheckoutSession
    let preview: ApiCheckoutPreview
}

struct ApiPaymentInput: Encodable {
    let type: String
    let paymentMethodId: String?
    let cpf: String?
}

struct ApiPaymentResult: Codable {
    let type: String
    let status: String
    let paymentMethodId: String?
    let displayName: String?
    let authorizationCode: String?
    let pixQrCodeBase64: String?
    let pixCopyPaste: String?
    let barcode: String?
    let digitableLine: String?
    let dueDate: String?
    let pdfUrl: String?
    let expiresAt: String?
}

struct ApiTicket: Codable {
    let ticketId: String
    let status: String
    let subject: String?
    let message: String?
    let orderId: String?
    let createdAt: String?
}

struct ApiFaqData: Codable {
    let topics: [ApiFaqItem]
}

// MARK: - Helpers de conversão

enum ApiMapping {
    private static let isoFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()

    private static let isoFormatterNoFraction = ISO8601DateFormatter()

    private static let displayFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "pt_BR")
        formatter.dateFormat = "dd/MM/yyyy"
        return formatter
    }()

    /// Converte string ISO8601 em data de exibição "dd/MM/yyyy"; se não for ISO, retorna como veio.
    static func displayDate(_ raw: String?) -> String {
        guard let raw, !raw.isEmpty else { return "" }
        if let date = isoFormatter.date(from: raw) ?? isoFormatterNoFraction.date(from: raw) {
            return displayFormatter.string(from: date)
        }
        return raw
    }

    static func orderStatus(_ raw: String) -> OrderStatus {
        switch raw.uppercased() {
        case "CONFIRMED": return .confirmed
        case "PREPARING": return .preparing
        case "SHIPPED": return .shipped
        case "DELIVERED", "RETURNED": return .delivered
        default: return .confirmed
        }
    }

    static func cardBrand(_ raw: String) -> CardBrand {
        CardBrand(rawValue: raw.lowercased()) ?? .unknown
    }

    static func couponType(_ raw: String) -> CouponType {
        raw.uppercased() == "PERCENT" ? .percent : .fixed
    }

    static func notificationType(_ raw: String) -> NotificationType {
        switch raw.uppercased() {
        case "ORDER": return .order
        case "PROMO": return .promo
        default: return .system
        }
    }

    static func paymentTypeString(_ type: CheckoutPaymentType) -> String {
        switch type {
        case .pix: return "PIX"
        case .card: return "CARD"
        case .boleto: return "BOLETO"
        }
    }
}

// MARK: - Mapeadores DTO → models de tela

extension ApiProduct {
    func toModel() -> Product {
        Product(
            id: id,
            name: name,
            imageURL: imageUrl,
            price: price,
            originalPrice: originalPrice ?? price,
            discountPercent: discountPercent ?? 0,
            rating: Float(rating),
            reviewCount: reviewCount,
            isFreeShipping: isFreeShipping,
            isExpress: isExpress,
            category: category,
            specs: specs ?? [:]
        )
    }
}

extension ApiCategory {
    func toModel() -> Category {
        Category(id: id, name: name, icon: icon, colorHex: colorHex)
    }
}

extension ApiHomeSection {
    func toModel() -> HomeSection {
        HomeSection(id: id, title: title, productIds: productIds)
    }
}

extension ApiUser {
    func toModel() -> User {
        User(
            name: name,
            email: email,
            phone: phone,
            avatarInitial: avatarInitial ?? String(name.prefix(1)).uppercased(),
            isPlus: isPlus ?? false
        )
    }
}

extension ApiAddress {
    func toModel() -> Address {
        Address(
            id: id,
            label: label,
            zipCode: zipCode,
            street: street,
            number: number,
            complement: complement ?? "",
            neighborhood: neighborhood,
            city: city,
            state: state,
            isDefault: isDefault
        )
    }
}

extension Address {
    func toInput() -> ApiAddressInput {
        ApiAddressInput(
            label: label,
            zipCode: zipCode,
            street: street,
            number: number,
            complement: complement.isEmpty ? nil : complement,
            neighborhood: neighborhood,
            city: city,
            state: state,
            isDefault: isDefault
        )
    }
}

extension ApiPaymentMethod {
    func toModel() -> PaymentMethod {
        PaymentMethod(
            id: id,
            brand: ApiMapping.cardBrand(brand),
            lastFour: lastFour,
            expiry: expiry,
            holderName: holderName,
            label: label,
            isDefault: isDefault
        )
    }
}

extension ApiShippingOption {
    func toModel() -> ShippingOption {
        ShippingOption(
            id: id,
            name: name,
            deliveryEstimate: deliveryEstimate,
            price: price,
            isExpress: isExpress ?? false
        )
    }
}

extension ApiCoupon {
    func toModel() -> Coupon {
        Coupon(
            code: code,
            description: description,
            type: ApiMapping.couponType(type),
            value: value,
            expiry: ApiMapping.displayDate(expiry)
        )
    }
}

extension ApiNotification {
    func toModel() -> AppNotification {
        AppNotification(
            id: id,
            type: ApiMapping.notificationType(type),
            title: title,
            body: body,
            date: ApiMapping.displayDate(date),
            isRead: isRead
        )
    }
}

extension ApiChatMessage {
    func toModel() -> ChatMessage {
        ChatMessage(id: id, text: text, isAgent: isAgent, time: time)
    }
}

extension ApiReview {
    func toModel() -> Review {
        Review(
            id: id,
            productId: productId,
            author: author,
            rating: rating,
            date: ApiMapping.displayDate(date),
            text: text,
            photoURLs: photoUrls ?? []
        )
    }
}

extension ApiFaqItem {
    func toModel() -> FaqItem {
        FaqItem(question: question, answer: answer)
    }
}

extension ApiTicket {
    func toModel() -> SupportTicket {
        SupportTicket(
            ticketId: ticketId,
            status: status,
            subject: subject ?? "",
            message: message ?? "",
            orderId: orderId,
            createdAt: ApiMapping.displayDate(createdAt)
        )
    }
}

extension ApiCart {
    /// Resolve os itens em `CartItem`, usando o produto embutido ou o catálogo local.
    func toModels(catalog: [Product]) -> [CartItem] {
        items.compactMap { item in
            let product = item.product?.toModel() ?? catalog.first { $0.id == item.productId }
            guard let product else { return nil }
            return CartItem(product: product, quantity: item.quantity)
        }
    }
}

extension ApiOrderPaymentMethod {
    func toModel() -> PaymentMethod {
        PaymentMethod(
            id: type.lowercased(),
            brand: .unknown,
            lastFour: "",
            expiry: "",
            holderName: displayName,
            label: label ?? displayName,
            isDefault: false
        )
    }
}

extension ApiOrder {
    func toModel(catalog: [Product]) -> Order {
        let cartItems: [CartItem] = items.compactMap { item in
            let product = item.product?.toModel() ?? catalog.first { $0.id == item.productId }
            guard let product else { return nil }
            return CartItem(product: product, quantity: item.quantity)
        }
        let history = (statusHistory ?? []).map {
            OrderStatusEntry(
                status: ApiMapping.orderStatus($0.status),
                date: ApiMapping.displayDate($0.date),
                location: $0.location ?? ""
            )
        }
        return Order(
            id: id,
            items: cartItems,
            total: total,
            status: ApiMapping.orderStatus(status),
            deliveryDate: deliveryDate ?? "",
            address: address?.toModel(),
            paymentMethod: paymentMethod?.toModel(),
            subtotal: subtotal,
            shipping: shipping,
            discount: discount + (pixDiscount ?? 0),
            trackingCode: trackingCode ?? "",
            statusHistory: history
        )
    }
}
